### PR TITLE
WIP: TSDB: Make into service

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
Make the TSDB subsystem into a dependency injected service (`tsdb.Service`). Just a draft to elicit feedback!

I'm open to providing a condensed overview verbally, if the changes are just too much to take in. A huge part of the changes were made in order to eliminate dependency cycles between packages, that arose when moving logic/state into `tsdb.Service`.

If the direction of the PR is agreed on, I can extract smaller PRs.

TODO:

- [x] Fix Postgres data source
- [x] Fix TODOs in `DataSourcePlugin`
- [x] Fix TODO in `FrontendPluginBase`
- [x] Fix TODO in `plugins.GetDataPlugin`
- [ ] Fix tests
- [ ] Consider if updateAppDashboards still works after moving to `plugindashboards.Service`
- [ ] Test ngalert after changes
- [ ] Test cloud data sources after changes

### CODE REVIEW

**Review by torkelo — COMMENTED** [review]
Looks good, so this moves a lot of types under tsdb/models to plugins/models , understand that this was to avoid cyclic dependencies , but how come those cyclic dependencies manifiest now, would they not have been present when accessing tsdb.HandleRequest as well? 

Or was it the use of the bus that hid the dependency? 

Can you show a drawing (google slide boxes or something) of the dependency graph? can help to visualize it to understand the root problem and where thing should live. 

My only real concern is the prefixing of a lot of types with TSDB, both from a readability standpoint (4 uppercase letters in a type I think we should avoid), and that these are for more than time series (old naming/use case), in frontend we just call these DataQuery , DataQueryReponse etc

**Comment by aknuds1** [comment]
> Looks good, so this moves a lot of types under tsdb/models to plugins/models , understand that this was to avoid cyclic dependencies , but how come those cyclic dependencies manifiest now, would they not have been present when accessing tsdb.HandleRequest as well?
>
> Or was it the use of the bus that hid the dependency?

It's very difficult to say in retrospect exactly what all the different cycles look like, but at least some of them were probably hidden through the use of dynamic dispatch (the bus).
 
> Can you show a drawing (google slide boxes or something) of the dependency graph? can help to visualize it to understand the root problem and where thing should live.

I'm not sure if a drawing is necessary. The one cycle I remember well is basically from the tsdb package to the plugins package and back, i.e. tsdb => plugins => tsdb. There might have been some indirection between also, but what I do recall code wise is that the TSDB subsystem would use plugins and plugins would in turn use TSDB. Although it would compile before, it's pretty mind blowing to try to reason about how that code flows at runtime.

It will be very difficult to analyze dependency cycles in more depth than this, because I only discovered them through fixing compilation failures over a week of work or so. It's a process I'd prefer not to repeat :/

> My only real concern is the prefixing of a lot of types with TSDB, both from a readability standpoint (4 uppercase letters in a type I think we should avoid), and that these are for more than time series (old naming/use case), in frontend we just call these DataQuery , DataQueryReponse etc

How do you suggest we solve this? Can you f.ex. make a branch of mine and make renaming changes there? I named the best I could, based on these symbols (types, functions) looking as if they were associated with TSDB.

File: pkg/plugins/plugindashboards/service.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+func init() {
+	registry.Register(&registry.Descriptor{
+		Name:     "PluginDashboardService",
+		Instance: &Service{},
+	})
+}
+
+type Service struct {
+	TSDBService   *tsdb.Service          `inject:""`
```
Comment (torkelo) [inline pkg/plugins/plugindashboards/service.go:23]: Wander what would be a better name for this service, DataRequestService?

File: pkg/plugins/models/tsdb.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+	RefID         string             `json:"refId"`
+	Model         *simplejson.Json   `json:"model,omitempty"`
+	DataSource    *models.DataSource `json:"datasource"`
+	MaxDataPoints int64              `json:"maxDataPoints"`
+	IntervalMS    int64              `json:"intervalMs"`
+	QueryType     string             `json:"queryType"`
+}
+
+// TSDBQuery contains all information about a TSDB query request.
+type TSDBQuery struct {
```
Comment (torkelo) [inline pkg/plugins/models/tsdb.go:26]: Would try to remove all the TSDB prefixes from these types and replace them with Data

Comment (aknuds1) [reply pkg/plugins/models/tsdb.go]: Done.

File: pkg/plugins/plugindashboards/service.go (hunk 1 above)
Comment (aknuds1) [reply pkg/plugins/plugindashboards/service.go]: Done.

**Pull request closed without merging**